### PR TITLE
Correctly display DCA warnings and fix some incorrect warnings that showed up

### PR DESCRIPTION
**Pull request by mohammadfawaz**
Closes https://github.com/FuelLabs/sway/issues/2143
Closes https://github.com/FuelLabs/sway/issues/306

The missing `!` in DCA was hiding some warnings. When added, a bunch of warnings showed up from `lib-core` and `lib-std`:
- One in `sway-lib-core` due to a duplicate declaration of `OrdEq` (likely a git merge issue?)
- A few others related to unused constant declarations:
  - Some of those need to be `pub` so I enabled that keyword for constants and fixed them.
  - Some don't need to be `pub` but still emit a warning. Turns out `asm` blocks needed some additional handling in DCA.
    - Those that don't need `pub` still caused an error in IR when the library code that uses them is imported. The fix is to remove the visibility checking in the const expr evaluator because the actual check happens earlier in type checking anyways and the existing check is causing the const not to be available to the library code itself.

Another instance of incorrect warnings showed up with `storage` declarations which I also fixed.

### CODE REVIEW

File: sway-core/src/control_flow_analysis/dead_code_analysis.rs [hunk 1]
```diff
@@ -148,53 +148,70 @@ impl ControlFlowGraph {
...
+                            // Don't emit warnings for constant declarations in libraries. We
+                            // cannot yet figure out whether they're actually used outside the
+                            // library or not.
+                            ControlFlowGraphNode::ProgramNode(TypedAstNode {
+                                content:
+                                    TypedAstNodeContent::Declaration(
+                                        TypedDeclaration::ConstantDeclaration(_),
+                                    ),
+                                ..
+                            }) => true,
```
Comment (mohammadfawaz) [inline sway-core/src/control_flow_analysis/dead_code_analysis.rs:209]: This is the only actual addition to this file. Everything else is due to the formatter.

File: sway-lib-core/src/ops.sw [hunk 2]
```diff
@@ -459,16 +459,6 @@ trait OrdEq: Ord + Eq {
     }
 }
 
-trait OrdEq: Ord + Eq {
```
Comment (mohammadfawaz) [inline sway-lib-core/src/ops.sw:462]: Duplicate trait declaration.

File: sway-core/src/control_flow_analysis/dead_code_analysis.rs [hunk 3]
```diff
@@ -100,7 +95,7 @@ impl ControlFlowGraph {
             .filter(|CompileWarning { span, .. }| {
                 // if any other warnings contain a span which completely covers this one, filter
                 // out this one.
-                all_warnings.iter().any(
+                !all_warnings.iter().any(
```
Comment (mohammadfawaz) [inline sway-core/src/control_flow_analysis/dead_code_analysis.rs:98]: This is the main fix.

File: sway-lib-std/src/assert.sw [hunk 4]
```diff
@@ -3,7 +3,7 @@ library assert;
 use ::logging::log;
 use ::revert::revert;
 
-const FAILED_REQUIRE_SIGNAL = 42;
+pub const FAILED_REQUIRE_SIGNAL = 42;
```
Comment (mohammadfawaz) [inline sway-lib-std/src/assert.sw:6]: This shouldn't really need the `pub` keyword but without it I get an IR error. Will create an issue.

File: sway-lib-core/src/ops.sw (hunk 2 above)
Comment (mohammadfawaz) [inline sway-lib-core/src/ops.sw:462]: Duplicate trait declaration.

File: sway-core/src/ir_generation/const_eval.rs [hunk 5]
```diff
@@ -47,15 +46,8 @@ pub fn compile_const_decl(
                 TypedDeclaration::ConstantDeclaration(TypedConstantDeclaration {
                     name,
                     value,
-                    visibility,
-                }) => {
-                    // XXX Do we really only add public constants?
```
Comment (otrho) [inline sway-core/src/ir_generation/const_eval.rs:52]: This was a comment from when I copied the older AST->ASM code which only compiled pubs, and it has languished here for 6 months!  Thanks. 🙂

**Review by otrho — APPROVED** [review]

**Review by mitchmindtree — APPROVED** [review]
Lgtm!

**Pull request merged**